### PR TITLE
[Core, Helper] Add spelling suggestion in ObjectFactory

### PR DESCRIPTION
**Pull request by damienmarchal**
In this PR we:
- add difflib in extlibs, difflib is an header only library that allows to make approximated search for string. 
- add sofa::helper::DiffLib to wrap the header only into a sofa looking library. 
- use DiffLib to implement spelling suggestion when a component cannot be created by the ObjectFactory.
Example of output:
```console
[ERROR]   [SofaPython3::SceneLoader] Unable to completely load the scene from file 'toto.py'.  
Python exception:  ValueError: Object type MechanicaObject<> was not created  
The object is not in the factory.  
But the following exits:  
                      : MechanicalObject  (98% match)
                      : MappedObject  (54% match)
                      : MechanicalStateController  (53% match)
```
 
The DiffLib feature is extensively used to implement a much wider spelling suggestion mechanism in SofaPython3.
Have a look at all the examples and use cases in PR  https://github.com/sofa-framework/SofaPython3/pull/314. 

I'm not using xml scenes so I didn't implemented support for spelling suggestion there. If your are interested in doing such a thing, the SofaPython3 implementation is a good starting point for your PR. 

___________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by damienmarchal** [comment]
[ci-build][with-all-tests]

File: extlibs/CMakeLists.txt [hunk 1]
```diff
@@ -10,6 +10,9 @@ set(EXTLIBS_TARGETS
 set(JSON_ROOT "${CMAKE_CURRENT_LIST_DIR}/json" CACHE PATH "Nlohmann JSON directory")
 # STB (header-only)
 set(STB_ROOT "${CMAKE_CURRENT_LIST_DIR}/stb" CACHE PATH "STB directory")
+# DiffLib
+# STB (header-only)
```
Comment (alxbilger) [inline extlibs/CMakeLists.txt]:
```suggestion
```

File: Sofa/framework/Core/src/sofa/core/objectmodel/Base.cpp [hunk 2]
```diff
@@ -420,7 +423,19 @@ bool Base::parseField( const std::string& attribute, const std::string& value)
     std::vector< BaseLink* > linkVec = findLinks(attribute);
     if (dataVec.empty() && linkVec.empty())
     {
-        msg_warning() << "Unknown Data field or Link: " << attribute ;
+        std::vector<std::string> possibleNames;
```
Comment (alxbilger) [inline Sofa/framework/Core/src/sofa/core/objectmodel/Base.cpp:426]:
```suggestion
        std::vector<std::string> possibleNames;
        possibleNames.reserve(m_vecData.size() + m_vecLink.size());
```

File: Sofa/framework/Core/src/sofa/core/ObjectFactory.cpp [hunk 3]
```diff
@@ -207,12 +208,26 @@ objectmodel::BaseObject::SPtr ObjectFactory::createObject(objectmodel::BaseConte
         using sofa::helper::lifecycle::uncreatableComponents;
         if(it == registry.end())
         {
-            arg->logError("The object is not in the factory.");
+            arg->logError("The object '"+classname+"' is not in the factory.");
```
Comment (alxbilger) [inline Sofa/framework/Core/src/sofa/core/ObjectFactory.cpp]:
```suggestion
            arg->logError("The object '" + classname + "' is not in the factory.");
```

File: Sofa/framework/Core/src/sofa/core/ObjectFactory.cpp [hunk 4]
```diff
@@ -207,12 +208,26 @@ objectmodel::BaseObject::SPtr ObjectFactory::createObject(objectmodel::BaseConte
...
             auto uuncreatableComponent = uncreatableComponents.find(classname);
             if( uuncreatableComponent != uncreatableComponents.end() )
             {
                 arg->logError( uuncreatableComponent->second.getMessage() );
             }
+            else
+            {
+                std::vector<std::string> possibleNames;
```
Comment (alxbilger) [inline Sofa/framework/Core/src/sofa/core/ObjectFactory.cpp:219]:
```suggestion
                std::vector<std::string> possibleNames;
                possibleNames.reserve(registry.size());
```

**Review by alxbilger — CHANGES_REQUESTED** [review]
Almost good to go.

It worth mentioning that the difflib comes from https://github.com/duckie/difflib.

File: extlibs/difflib/difflib.h [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+// all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
```
Comment (damienmarchal) [inline extlibs/difflib/difflib.h]:
```suggestion
/// Original source: https://github.com/duckie/difflib.
```

**Review by alxbilger — APPROVED** [review]

**Comment by alxbilger** [comment]
@damienmarchal Is the work finished on this PR?

**Comment by damienmarchal** [comment]
@alxbilger  yes, I have no more time to improve it. 

There is still work to do in order to generalize the usage of that everywhere we are doing "name" search, but this is future work.

**Comment by damienmarchal** [comment]
Ready... so merge.

**Pull request merged**